### ISSUE TITLE
Output file not written when asserts fail

### ISSUE THREAD

**Issue by lambrospetrou** [Orange-OpenSource/hurl#3158]
### What is the current *bug* behavior?

When there is an `output` option specified for an entry but the asserts fail, even though the request was successfully completed and we have a response body, then the output file is not written.

The code right now checks if there were any errors, including asserts, before writing the output file ([see output write code](https://github.com/Orange-OpenSource/hurl/blob/master/packages/hurl/src/runner/hurl_file.rs#L354) and [see asserts becoming errors code](https://github.com/Orange-OpenSource/hurl/blob/master/packages/hurl/src/runner/entry.rs#L178)).

This is an issue for Skybear.NET platform, since using `output:` option is the only way to debug really what happens, and users specifying that option will expect that they will have the file with the response so that they can see how to adjust their scripts just through the UI.

Even locally, I would like to have the response so that I can see the full thing and fix my asserts.

### Steps to reproduce

```hurl
# test-output-with-assert-error.hurl
GET https://about.skybear.net/_live-demo/sample-001.json
[Options]
output: call1.json
HTTP 200
[Asserts]
header "content-type" endsWith "/json2"
```

Run the above:
```sh
hurl --json --test test-output-with-assert-error.hurl
```

### What is the expected *correct* behavior?

I expected to see the file `call1.json` to contain the response of the request, since the request completed without issues.

When the assert is correct, the file is output just fine.

### Execution context

- Hurl Version (`hurl --version`):

```
hurl 4.3.0 (x86_64-pc-linux-gnu) libcurl/7.81.0 OpenSSL/3.0.2 zlib/1.2.11 brotli/1.0.9 zstd/1.4.8 libidn2/2.3.2 libssh/0.9.6/openssl/zlib nghttp2/1.43.0
Features (libcurl):  alt-svc AsynchDNS brotli HSTS HTTP2 IDN IPv6 Largefile libz NTLM NTLM_WB SPNEGO SSL TLS-SRP UnixSockets zstd
Features (built-in): brotli
```

### Possible fixes

 
The explicitly specified output files should always be created even if asserts failed, since the request succeeded (e.g. no DNS error, no fetching error, no HTTP connection error).

Maybe differentiate `errors` ([see code](https://github.com/Orange-OpenSource/hurl/blob/master/packages/hurl/src/runner/entry.rs#L178)) into `assertErrors` and `httpConnectionErrors`, and only ignore the output file for the latter.

**Comment by jcamiel** [comment]
Hi @lambrospetrou 

We try to "work like" curl as much as possible. When there are network or protocol error, `curl` doesn't provide any body (obviously, since there are no body to provide). For the moment, we don't discriminate network/protocol errors and assert errors. For instance with [`--retry`](https://hurl.dev/docs/manual.html#retry)

```hurl
GET https://foo.com
[Options]
retry: 10
HTTP 200
```

Will retry request in the presence of network error, or HTTP status code != 200.
So in presence of error (network OR assert), we don't write response body.

I'm reluctant to change this: one could rely on the response being written after every check has passed (despite exit code). It changes for instance the behaviour of such a code:

```hurl
GET https://foo.com
[Options]
retry: 10
output: foo.bin
HTTP 200
```

In current Hurl, `foo.bin` is written only once on success, or never.

Now, there is the use case of getting the HTTP response even in the presence of assert error, a feature that helps tremendously for debugging. I see some options:

- `--error-format long`: we have created this option precisely for getting the HTTP response on assert errors. This option outputs a more verbose error on standard error, with the response body. The cons is that the log is not structured and can be difficult to process outside CLI. 
- `--report-json`: this report should have every aspect of a test result, included failed response body. I've the impression that with this option, one can get, in a stuctured way, every data from a call.
- we have also this issue #2499 : provide error in JSON. This is needed for a better integration with IDE but other use case like <https://skybear.net> could benefit from it. If is is feasible, we can imagine provide HTTP datas in this JSON (body, headers, etc...)

**Comment by lambrospetrou** [comment]
> --report-json: this report should have every aspect of a test result, included failed response body. I've the impression that with this option, one can get, in a stuctured way, every data from a call.

This should be enough, and the most wanted option really to me :) The `--error-format=json` can also work, but between these two I prefer the report which is complete.

> In current Hurl, foo.bin is written only once on success, or never.

I feel that this is not a breaking change, unless some Hurl users check the "file existence" as an all tests passed guarantee, which is weird. Although, I can understand you wanting not to change that behavior regardless...

**Comment by jcamiel** [comment]
Yes, `--report-json` will be included in 5.0.0 (we'll release it in the next few days). I've tested it and we have response body / headers even in case of asserts error.

**Comment by lambrospetrou** [comment]
I confirm that in 5.0.1 this works great BTW. The `--report-json` option does everything nicely. We have the response headers, and we have the response body even when a request failed.
Already used on #SkybearNET :)
